File: 198-house-robber/house-robber.py
class Solution(object):
    def rob(self, nums):
        rob1, rob2 = 0, 0
        #[rob1, rob2, n, n+1, ...]
        for cur_val in nums:
            temp=max(cur_val+rob1, rob2)
            rob1=rob2
            rob2 = temp
        return rob2


      
       
       

       
        

File: 198-house-robber/house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        vector<int> dp(nums.size(),-1);
        int n=nums.size();
        if(n == 1) return nums[0];
        dp[n-1]=nums[n-1];
        dp[n-2]=max(nums[n-1], nums[n-2]);
        for(int i=n-3; i>=0; i--){
            int rob=nums[i]+dp[i+2];
            int skip=dp[i+1];
            dp[i]=max(rob, skip);
        }
        return dp[0];
        
    }
   
};